src/action/list_directory.rs

```diff
@@ -0,0 +1,730 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the list directory action.
+//!
+//! A list directory action stats all files in the provided directory.
+
+use crate::session::{self, Session};
+use rrg_proto::{ListDirRequest, StatEntry, path_spec::PathType,
+                path_spec::Options, micros};
+
+use std::fs::{self, Metadata};
+use std::path::{PathBuf, Path};
+use std::fmt::{Display, Formatter};
+use log::warn;
+use std::time::SystemTime;
+
+/// An error type used when some path can't be read. E.g. listed directory
+/// path or some inner file path.
+#[derive(Debug)]
+enum Error {
+    ReadPath(std::io::Error),
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => Some(error),
+        }
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => {
+                write!(fmt, "unable to read path: {}", error)
+            }
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+/// Contains information about an `UnsupportedValue` error.
+#[derive(Debug)]
+struct UnsupportedValueError {
+    field: String,
+    value: String,
+}
+
+impl ParseError {
+
+    /// Constructs an `UnsupportedValue` from provided `field` and
+    /// `value`, then  converts it to `ParseError`.
+    fn unsupported_value(field: String, value: String) -> session::ParseError {
+        session::ParseError::malformed(ParseError::UnsupportedValue(
+            UnsupportedValueError {
+                field,
+                value,
+            }))
+    }
+}
+
+/// An error type used when the value which comes from protocol buffer is
+/// unsupported by the current client.
+#[derive(Debug)]
+enum ParseError {
+    UnsupportedValue(UnsupportedValueError),
+}
+
+impl std::error::Error for ParseError {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use ParseError::*;
+
+        match *self {
+            UnsupportedValue(_) => None,
+        }
+    }
+}
+
+impl Display for ParseError {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use ParseError::*;
+
+        match *self {
+            UnsupportedValue(ref message) => {
+                write!(fmt, "value {} in {} field is not supported",
+                       message.value, message.field)
+            }
+        }
+    }
+}
+
+impl From<ParseError> for session::Error {
+
+    fn from(error: ParseError) -> session::Error {
+        session::Error::action(error)
+    }
+}
```

Comment:
This should be implemented for `session::ParseError`, not action execution-specific `session::Error`.

Reply:
Fixed.